src/net_processing.cpp

```diff
@@ -3329,7 +3329,13 @@ bool ProcessMessage(CNode& pfrom, const std::string& msg_type, CDataStream& vRec
         pfrom.fSentAddr = true;
 
         pfrom.vAddrToSend.clear();
-        std::vector<CAddress> vAddr = connman->GetAddresses();
+        std::vector<CAddress> vAddr;
+        if (pfrom.HasPermission(PF_ADDR)) {
+            vAddr = connman->GetAddresses(nullopt);
+        } else {
+            Network peer_network = pfrom.addr.GetNetwork();
+            vAddr = connman->GetAddresses(peer_network);
+        }
         FastRandomContext insecure_rand;
         for (const CAddress &addr : vAddr) {
             if (!banman->IsBanned(addr)) {
```

Comment:
Perhaps this filtering out banned addresses shouldn't happen if we're sending from the cache. If an address in the cache is banned during the cache's lifetime, subsequent GETADDR requests will receive a different response, which could be a marginal privacy leak.

Reply:
In fact, I think this filtering out of banned addresses should move to `CConnman::GetAddresses()`. The only other client of that function is `getnodeaddresses` in rpc/net, and that probably doesn't want to know about banned addresses either.